File: MoonEngine-Root/Assets/src/main/java/com/greentree/commons/assets/serializator/IterableAssetSerializator.java
package com.greentree.commons.assets.serializator;

import com.greentree.commons.assets.key.AssetKey;
import com.greentree.commons.assets.key.IterableAssetKey;
import com.greentree.commons.assets.serializator.context.LoadContext;
import com.greentree.commons.assets.serializator.manager.CanLoadAssetManager;
import com.greentree.commons.assets.value.Value;
import com.greentree.commons.assets.value.merge.MIValue;
import com.greentree.commons.util.classes.info.TypeInfo;
import com.greentree.commons.util.iterator.IteratorUtil;

public final class IterableAssetSerializator<T> extends AbstractIterableAssetSerializator<T> {
	
	public IterableAssetSerializator(Class<T> cls) {
		super(cls);
	}
	
	public IterableAssetSerializator(TypeInfo<T> type) {
		super(type);
	}
	
	public Value<Iterable<T>> load(LoadContext context, AssetKey ckey) {
		if(ckey instanceof IterableAssetKey key) {
			final var iter = IteratorUtil.clone(IteratorUtil.map(key, k-> {
				return context.load(ITER_TYPE, k);
			}));
			
			return new MIValue<>(iter);
		}
		return null;
	}
	
	@Override
	public boolean canLoad(CanLoadAssetManager manager, AssetKey key) {
		if(key instanceof IterableAssetKey k) {
			for(var s : k)
				if(!manager.canLoad(ITER_TYPE, s))
					return false;
			return true;
		}
		return false;
	}
	
}

File: MoonEngine/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/IterableAssetSerializator.kt
package com.greentree.engine.moon.assets.serializator

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.engine.moon.assets.NotSupportedKeyType
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.key.IterableAssetKey
import com.greentree.engine.moon.assets.provider.AssetProvider
import com.greentree.engine.moon.assets.provider.MIAssetProvider
import com.greentree.engine.moon.assets.serializator.loader.AssetLoader

class IterableAssetSerializator<T : Any> : AbstractIterableAssetSerializator<T> {
	constructor(cls: Class<T>) : super(cls)
	constructor(type: TypeInfo<T>) : super(type)

	override fun load(manager: AssetLoader.Context, key: AssetKey): AssetProvider<Iterable<T>> {
		if(key is IterableAssetKey) {
			val iter = key.map { manager.load(ITER_TYPE, it) }
			return MIAssetProvider(iter)
		}
		throw NotSupportedKeyType
	}
}